### PR TITLE
Rework of ORM

### PR DESCRIPTION
**Pull request by davidprueser**
reworking orm structure to automatically join information and shrink size of user queries.

### CODE REVIEW

File: src/pycram/orm/base.py [hunk 1]
```diff
@@ -55,17 +70,18 @@ class MetaData(Base):
...
+    created_at: Mapped[datetime.datetime] = mapped_column(server_default=sqlalchemy.sql.functions.current_timestamp(),
+                                                          init=False)
     """The timestamp where this row got created. This is an aid for versioning."""
 
-    created_by = sqlalchemy.Column(sqlalchemy.String(255), default=os.getlogin())
+    created_by: Mapped[str] = mapped_column(default=os.getlogin(), init=False)
     """The user that created the experiment."""
 
-    description = sqlalchemy.Column(sqlalchemy.String(255), default=None, nullable=False)
+    description: Mapped[str] = mapped_column(init=False)
```
Comment (tomsch420) [inline src/pycram/orm/base.py:80]: Why is this init=False?

File: src/pycram/orm/base.py [hunk 2]
```diff
@@ -93,73 +109,77 @@ def reset(cls):
...
 
     __tablename__ = "Position"
 
-    x = sqlalchemy.Column(sqlalchemy.types.Float)
-    y = sqlalchemy.Column(sqlalchemy.types.Float)
-    z = sqlalchemy.Column(sqlalchemy.types.Float)
+    x: Mapped[float]
+    y: Mapped[float]
+    z: Mapped[float]
+    metadata_id: Mapped[Optional[int]] = mapped_column(ForeignKey("MetaData.id"), default=None)
```
Comment (tomsch420) [inline src/pycram/orm/base.py:120]: Why are these declared here?

File: src/pycram/orm/base.py [hunk 3]
```diff
@@ -93,73 +109,77 @@ def reset(cls):
...
     __tablename__ = "RobotState"
 
-    position = sqlalchemy.Column(sqlalchemy.types.Integer, sqlalchemy.ForeignKey("Position.id"))
+    position: Mapped[int] = mapped_column(ForeignKey("Position.id"), init=False)
+    position_table_entry: Mapped["Position"] = relationship(init=False)
     """The position of the robot."""
 
-    orientation = sqlalchemy.Column(sqlalchemy.types.Integer, sqlalchemy.ForeignKey("Quaternion.id"))
+    orientation: Mapped[int] = mapped_column(ForeignKey("Quaternion.id"), init=False)
+    orientation_table_entry: Mapped["Quaternion"] = relationship(init=False)
```
Comment (tomsch420) [inline src/pycram/orm/base.py:178]: The actual object quaternion should be in the quaternion field. The foreign key should have a name like quaternion_fk or something consistent.

File: src/pycram/orm/motion_designator.py [hunk 4]
```diff
@@ -55,11 +61,14 @@ class PickUpMotion(MotionDesignator):
...
     __tablename__ = "PickUpMotion"
-    id = sqlalchemy.Column(sqlalchemy.types.Integer, sqlalchemy.ForeignKey("Motion.id"), primary_key=True)
-    object = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey("Object.id"))
-    arm = sqlalchemy.Column(sqlalchemy.types.String(255))
-    gripper = sqlalchemy.Column(sqlalchemy.types.String(255))
-    grasp = sqlalchemy.Column(sqlalchemy.types.String(255))
+
+    id: Mapped[int] = mapped_column(ForeignKey("Motion.id"), primary_key=True, init=False)
+    object: Mapped[int] = mapped_column(ForeignKey("Object.id"), init=False)
+    object_table_entry: Mapped[ObjectDesignator] = relationship(init=False)
```
Comment (tomsch420) [inline src/pycram/orm/motion_designator.py:67]: naming

File: src/pycram/orm/motion_designator.py [hunk 5]
```diff
@@ -76,12 +85,17 @@ class PlaceMotion(MotionDesignator):
...
-    gripper = sqlalchemy.Column(sqlalchemy.types.String(255))
-    position = sqlalchemy.Column(sqlalchemy.types.Integer, sqlalchemy.ForeignKey("Position.id"))
-    orientation = sqlalchemy.Column(sqlalchemy.types.Integer, sqlalchemy.ForeignKey("Quaternion.id"))
+
+    id: Mapped[int] = mapped_column(ForeignKey("Motion.id"), primary_key=True, init=False)
+    object: Mapped[int] = mapped_column(ForeignKey("Object.id"), init=False)
+    object_table_entry: Mapped[ObjectDesignator] = relationship(init=False)
+    arm: Mapped[str] = mapped_column(init=False)
+    gripper: Mapped[str] = mapped_column(init=False)
+    position: Mapped[int] = mapped_column(ForeignKey("Position.id"), init=False)
```
Comment (tomsch420) [inline src/pycram/orm/motion_designator.py:94]: naming

File: src/pycram/orm/motion_designator.py [hunk 6]
```diff
@@ -98,13 +112,15 @@ class AccessingMotion(MotionDesignator):
...
-    part_of = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey("Object.id"))
-    arm = sqlalchemy.Column(sqlalchemy.types.String(255))
-    gripper = sqlalchemy.Column(sqlalchemy.types.String(255))
-    distance = sqlalchemy.Column(sqlalchemy.types.Float)
-    drawer_joint = sqlalchemy.Column(sqlalchemy.types.String(255))
-    drawer_handle = sqlalchemy.Column(sqlalchemy.types.String(255))
+
+    id: Mapped[int] = mapped_column(ForeignKey("Motion.id"), primary_key=True, init=False)
+    part_of: Mapped[int] = mapped_column(ForeignKey("Object.id"), init=False)
+    object_table_entry: Mapped[ObjectDesignator] = relationship(init=False)
```
Comment (tomsch420) [inline src/pycram/orm/motion_designator.py:118]: naming

File: src/pycram/orm/action_designator.py [hunk 7]
```diff
@@ -1,18 +1,22 @@
...
     The purpose of this class is to correctly map the inheritance from the action designator class into the database.
     Inheritance is implemented as Joined Table Inheritance (see https://docs.sqlalchemy.org/en/20/orm/inheritance.html)
     """
     __tablename__ = "Action"
-    dtype = sqlalchemy.Column(sqlalchemy.types.String(255))
-    robot_state = sqlalchemy.Column(sqlalchemy.types.Integer, sqlalchemy.ForeignKey("RobotState.id"))
+
+    dtype: Mapped[str] = mapped_column(init=False)
+    robot_state: Mapped[int] = mapped_column(ForeignKey("RobotState.id"), init=False)
+    robot_state_table_entry: Mapped[RobotState] = relationship(init=False)
```
Comment (tomsch420) [inline src/pycram/orm/action_designator.py:19]: naming

File: src/pycram/orm/action_designator.py [hunk 8]
```diff
@@ -1,18 +1,22 @@
...
 
     __mapper_args__ = {
         "polymorphic_identity": __tablename__,
```
Comment (tomsch420) [inline src/pycram/orm/action_designator.py:22]: can polymorphic identities be declared in supertype via sqlalchemy property like the metadata id?

File: src/pycram/orm/action_designator.py [hunk 9]
```diff
@@ -23,74 +27,85 @@ class Action(Base):
...
+    arm: Mapped[Arms] = mapped_column(default=None)
 
     __mapper_args__ = {
         "polymorphic_identity": __tablename__,
     }
 
-    def __init__(self, arm: str = None):
-        super().__init__()
-        self.arm = arm
+    # def __init__(self, arm: str = None):
```
Comment (tomsch420) [inline src/pycram/orm/action_designator.py:38]: commented code block

File: src/pycram/orm/action_designator.py [hunk 10]
```diff
@@ -23,74 +27,85 @@ class Action(Base):
...
     """ORM Class of pycram.designators.action_designator.NavigateAction."""
 
     __tablename__ = "Navigate"
-    id = sqlalchemy.Column(sqlalchemy.types.Integer, sqlalchemy.ForeignKey("Action.id"), primary_key=True)
-    position = sqlalchemy.Column(sqlalchemy.types.Integer, sqlalchemy.ForeignKey("Position.id", ))
-    orientation = sqlalchemy.Column(sqlalchemy.types.Integer, sqlalchemy.ForeignKey("Quaternion.id"))
+
+    id: Mapped[int] = mapped_column(ForeignKey("Action.id"), primary_key=True, init=False)
+    position: Mapped[int] = mapped_column(ForeignKey("Position.id", ), default=None)
+    position_table_entry: Mapped[Position] = relationship(init=False)
```
Comment (tomsch420) [inline src/pycram/orm/action_designator.py:50]: naming

File: src/pycram/orm/action_designator.py [hunk 11]
```diff
@@ -23,74 +27,85 @@ class Action(Base):
...
     __mapper_args__ = {
         "polymorphic_identity": __tablename__,
     }
 
-    def __init__(self, position: Optional[int] = None, orientation: Optional[int] = None):
-        super().__init__()
-        self.position = position
-        self.orientation = orientation
+    # def __init__(self, position: Optional[int] = None, orientation: Optional[int] = None):
+    #     super().__init__()
```
Comment (tomsch420) [inline src/pycram/orm/action_designator.py:59]: commented code block

File: src/pycram/orm/action_designator.py [hunk 12]
```diff
@@ -23,74 +27,85 @@ class Action(Base):
...
 
 class MoveTorsoAction(Action):
     """ORM Class of pycram.designators.action_designator.MoveTorsoAction."""
+
     __tablename__ = "MoveTorso"
-    id = sqlalchemy.Column(sqlalchemy.types.Integer, sqlalchemy.ForeignKey("Action.id"), primary_key=True)
-    position = sqlalchemy.Column(sqlalchemy.types.Float)
+
+    id: Mapped[int] = mapped_column(ForeignKey("Action.id"), primary_key=True, init=False)
+    position: Mapped[Optional[float]] = mapped_column(default=None)
```
Comment (tomsch420) [inline src/pycram/orm/action_designator.py:70]: name is weird. This shoud be renamed to height

File: src/pycram/orm/action_designator.py [hunk 13]
```diff
@@ -23,74 +27,85 @@ class Action(Base):
...
 
     __mapper_args__ = {
         "polymorphic_identity": __tablename__,
     }
 
-    def __init__(self, position: Optional[float] = None):
-        super(MoveTorsoAction, self).__init__()
-        self.position = position
+    # def __init__(self, position: Optional[float] = None):
+    #     super(MoveTorsoAction, self).__init__()
```
Comment (tomsch420) [inline src/pycram/orm/action_designator.py:77]: commented 👏 code 👏 block 👏

File: src/pycram/orm/action_designator.py [hunk 14]
```diff
@@ -23,74 +27,85 @@ class Action(Base):
...
         "polymorphic_identity": __tablename__,
     }
 
-    def __init__(self, gripper: str, motion: str):
-        super(SetGripperAction, self).__init__()
-        self.gripper = gripper
-        self.motion = motion
+    # def __init__(self, gripper: str, motion: str):
+    #     super(SetGripperAction, self).__init__()
+    #     self.gripper = gripper
```
Comment (tomsch420) [inline src/pycram/orm/action_designator.py:96]: commented 👏 code 👏 block 👏

File: src/pycram/orm/action_designator.py [hunk 15]
```diff
@@ -23,74 +27,85 @@ class Action(Base):
...
+
     __tablename__ = "Release"
-    id = sqlalchemy.Column(sqlalchemy.types.Integer, sqlalchemy.ForeignKey("Action.id"), primary_key=True)
-    gripper = sqlalchemy.Column(sqlalchemy.types.String(255), nullable=False)
-    object = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey("Object.id"))
+
+    id: Mapped[int] = mapped_column(ForeignKey("Action.id"), primary_key=True, init=False)
+    gripper: Mapped[str] = mapped_column(init=False)
+    object: Mapped[int] = mapped_column(ForeignKey("Object.id"), init=False)
+    object_table_entry: Mapped[ObjectDesignator] = relationship(init=False)
```
Comment (tomsch420) [inline src/pycram/orm/action_designator.py:108]: naming

File: src/pycram/orm/action_designator.py [hunk 16]
```diff
@@ -99,11 +114,14 @@ class Release(Action):
...
-    id = sqlalchemy.Column(sqlalchemy.types.Integer, sqlalchemy.ForeignKey("Action.id"), primary_key=True)
-    gripper = sqlalchemy.Column(sqlalchemy.types.String(255), nullable=False)
-    effort = sqlalchemy.Column(sqlalchemy.types.Float, nullable=False)
-    object = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey("Object.id"))
+
+    id: Mapped[int] = mapped_column(ForeignKey("Action.id"), primary_key=True, init=False)
+    gripper: Mapped[str] = mapped_column(init=False)
+    effort: Mapped[float] = mapped_column(init=False)
+    object: Mapped[int] = mapped_column(ForeignKey("Object.id"), init=False)
+    object_table_entry: Mapped[ObjectDesignator] = relationship(init=False)
```
Comment (tomsch420) [inline src/pycram/orm/action_designator.py:124]: n 👏 a 👏 m 👏 i 👏 n 👏 g 👏

**Review by tomsch420 — CHANGES_REQUESTED** [review]
Also try to set the __tablename__ attribute in the Base class

File: src/pycram/orm/task.py [hunk 17]
```diff
@@ -1,42 +1,50 @@
...
-class Code(Base):
+
+    id: Mapped[int] = mapped_column(autoincrement=True, primary_key=True, init=False)
+
+    code: Mapped[int] = mapped_column(ForeignKey("Code.id"), default=None)
+    code_table_entry: Mapped["Code"] = relationship(init=False)
+    start_time: Mapped[datetime.datetime] = mapped_column(default=None)
+    end_time: Mapped[Optional[datetime.datetime]] = mapped_column(default=None)
+    status: Mapped[TaskStatus] = mapped_column(default=None)
+    reason: Mapped[Optional[str]] = mapped_column(default=None)
```
[the remark this answers is not in the document]
Comment (davidprueser) [reply src/pycram/orm/task.py:23]: What do you mean?

File: src/pycram/designators/object_designator.py [hunk 18]
```diff
@@ -50,14 +48,16 @@ def to_sql(self) -> ORMObjectPart:
...
+            obj.process_metadata_id = metadata.id
+            pose = self.part_pose.insert(session)
+            obj.pose_id = pose.id
             # try to create the part_of object
-            if self.part_of:
-                part = self.part_of.insert(session)
-                obj.part_of = part.id
-            else:
-                obj.part_of = None
+            # if self.part_of:
```
Comment (tomsch420) [inline src/pycram/designators/object_designator.py:56]: delete commented code blocks

**Review by tomsch420 — APPROVED** [review]

**Review by tomsch420 — APPROVED** [review]

**Pull request merged**